constraint-solver/tests/solver.rs

```diff
@@ -222,10 +222,91 @@ fn xor() {
     };
 
     let solver = Solver::new(constraint_system)
-        .with_bus_interaction_handler(Box::new(TestBusInteractionHandler {}));
+        .with_bus_interaction_handler(Rc::new(TestBusInteractionHandler {}));
 
     assert_solve_result(
         solver,
         vec![("a", 0xa0.into()), ("b", 0x0b.into()), ("c", 0xab.into())],
     );
 }
+
+#[test]
+fn one_hot_flags() {
+    let constraint_system = ConstraintSystem {
+        algebraic_constraints: vec![
+            // Boolean flags
+            var("flag0") * (var("flag0") - constant(1)),
+            var("flag1") * (var("flag1") - constant(1)),
+            var("flag2") * (var("flag2") - constant(1)),
+            var("flag3") * (var("flag3") - constant(1)),
+            // Exactly one flag is active
+            var("flag0") + var("flag1") + var("flag2") + var("flag3") - constant(1),
+            // Flag 2 is active
+            var("flag0") * constant(12)
+                + var("flag1") * constant(123)
+                + var("flag2") * constant(1234)
+                + var("flag3") * constant(12345)
+                - constant(1234),
+        ],
+        bus_interactions: vec![],
+    };
+
+    // This can be solved via backtracking:
+    // For any given flag, setting it to 1 will set the other flags to 0.
+    // For any flag except flag2, this will lead to a contradiction in the last constraint.
+    // So, the solver can just set the flags one-by-one, finding a unique assignment in each step.
+    assert_solve_result(
+        Solver::new(constraint_system),
+        vec![
+            ("flag0", 0.into()),
+            ("flag1", 0.into()),
+            ("flag2", 1.into()),
+            ("flag3", 0.into()),
+        ],
+    );
+}
+
+#[test]
+fn binary_flags() {
```

Comment:
This test is actually somewhat similar to the OpenVM loadstore flags:
https://github.com/openvm-org/openvm/blob/v1.0.0/extensions/rv32im/circuit/src/loadstore/core.rs#L113-L139

I think we should be able to solve this as well and the only thing that's missing is a pattern matcher for a `[0, 2]` range constraint!